api/desecapi/serializers.py

```diff
@@ -301,34 +301,72 @@ def to_internal_value(self, data):
+        As both "known"/"unknown" and "empty"/"non-empty" are binary partitions on `everything`, the combination of
+        both partitions `everything` in four disjoint subsets. Hence, every RRset in `everything` is taken care of.
+
+                   empty   |  non-empty
+        ------- | -------- | -----------
+        known   |  delete  |   update
+        unknown |  no-op   |   create
+
+        :param instance: List of relevant RRset objects, i.e. the Django.Model subclass instances. Relevant are all
+        instances that are referenced in `validated_data`. If a referenced RRset is missing from instances, it will be
+        considered unknown and hence be created. This may cause a database integrity error. If an RRset is given, but
+        not relevant (i.e. not referred to by `validated_data`), a ValueError will be raised.
+        :param validated_data: List of RRset data objects, i.e. dictionaries.
+        :return: List of RRset objects (Django.Model subclass) that have been created or updated.
+        """
+        def key(data_item):
+            return data_item.get('subname', ''), data_item['type']
+
+        def is_empty(data_item):
+            return data_item.get('records', None) == []
+
         instance_index = {(rrset.subname, rrset.type): rrset for rrset in instance}
-        data_index = {(data.get('subname', ''), data['type']): data for data in validated_data}
-
-        created = data_index.keys() - instance_index.keys()
-        updated = {(d.get('subname', ''), d['type']) for d in validated_data if d.get('records', None) != []} - created
-        deleted = {(d.get('subname', ''), d['type']) for d in validated_data if d.get('records', None) == []}
-
-        assert created | updated | deleted == data_index.keys(), 'Given data not covered. Expected %s, but got %s ' \
-                                                                 'and %s and %s.' % (data_index.keys(), created,
-                                                                                     updated, deleted)
-        assert not created & updated and not updated & deleted and not deleted & created, \
-            'Created, updated, deleted not pairwise disjoint.'
-        assert data_index.keys() | instance_index.keys() == data_index.keys(), \
-            'Instances is not a subset of data.'
-
-        # Note that for the following create and update actions, the data passed along is treated as validated,
-        # no further validation is conducted!
+        data_index = {key(data): data for data in validated_data}
+
+        if data_index.keys() | instance_index.keys() != data_index.keys():
+            raise ValueError('Given set of known RRsets (`instance`) is not a subset of RRsets referred to in'
+                             '`validated_data`. While this would produce the desired result, this is illegal due to its'
+                             ' inefficiency.')
+
+        everything = instance_index.keys() | data_index.keys()
```

Comment:
We have just checked that `data_index.keys() | instance_index.keys() == data_index.keys()`, i.e. we can just set `everything = data_index.keys()`.

Reply:
It's true, but if you remove the ValueError it's necessary to do it the way it's implemented.

Reply:
Ah, got it. For clarity, it would then be good to move this after the line that defines `data_index` and use `everything` in the `if` clause that leads to the `ValueError`. I think I should change my profession to nit-picking.